### PR TITLE
Add NTP exporter

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/11650

### CODE REVIEW

File: helm/net-exporter/templates/daemonset.yaml [hunk 1]
```diff
@@ -59,6 +59,9 @@ spec:
         {{ if (.Values.NetExporter.Hosts) }}
         - "-hosts={{ .Values.NetExporter.Hosts }}"
         {{ end }}
+        {{ if (.Values.NetExporter.NTPServers) }}
+        - "-ntp-servers={{ .Values.NetExporter.NTPServers }}"
```
Comment (kopiczko) [inline helm/net-exporter/templates/daemonset.yaml:63]: Do we have a case in which we need to specify custom NTP servers? If not I'd drop that flag.

File: main.go [hunk 2]
```diff
@@ -33,6 +35,7 @@ func init() {
 	flag.BoolVar(&disableDNSTCPCheck, "disable-dns-tcp-check", false, "Disable DNS TCP check")
 	flag.StringVar(&hosts, "hosts", "giantswarm.io.,kubernetes.default.svc.cluster.local.", "DNS hosts to resolve")
 	flag.StringVar(&namespace, "namespace", "monitoring", "Namespace of net-exporter service")
+	flag.StringVar(&ntpServers, "ntp-servers", "0.flatcar.pool.ntp.org,1.flatcar.pool.ntp.org", "NTP servers to use for time synchronization")
```
Comment (kopiczko) [inline main.go:38]: Let's use `StringSliceVar` instead.

File: main.go [hunk 3]
```diff
@@ -121,12 +124,29 @@ func main() {
...
+
+		c := ntp.Config{
+			Logger: logger,
+
+			NTPServers: splitNTPServers,
+		}
+
+		ntpCollector, err = ntp.New(c)
+		if err != nil {
+			panic(fmt.Sprintf("%#v\n", err))
```
Comment (kopiczko) [inline main.go:139]:
```suggestion
			panic(microerror.JSON(err))
```

File: ntp/ntp.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+	syncErrorCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: prometheus.BuildFQName(namespace, "", "sync_error_total"),
+			Help: "Total number of errors ntp syncs.",
+		},
+		[]string{"server"},
+	)
+
+	prometheus.MustRegister(errorCount)
+	prometheus.MustRegister(syncErrorCount)
```
Comment (kopiczko) [inline ntp/ntp.go:78]: Metrics are registered globally in prometheus so I'd rather do all of that in `init` function.

File: ntp/ntp.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+		ntpServers: config.NTPServers,
+
+		latencyHistogramVec: latencyHistogramVec,
+		latencyHistogramDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "latency_seconds"),
+			"Histogram of latency of NTP sync requests.",
+			[]string{"server"},
+			nil,
+		),
```
Comment (kopiczko) [inline ntp/ntp.go:91]: We usually define descs as package toplevel variables.

File: ntp/ntp.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+func (c *Collector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.latencyHistogramDesc
+}
+
+func (c *Collector) ntpsync(ntpServer string, latencyHistogramVec *histogramvec.HistogramVec) {
+	start := time.Now()
+
+	_, err := ntp.Time(ntpServer)
+	if err != nil {
+		c.logger.Log("level", "error", "message", fmt.Sprintf("could not sync time with ntp server %#q", ntpServer), "stack", microerror.JSON(err))
```
Comment (kopiczko) [inline ntp/ntp.go:110]:
```suggestion
		c.logger.Log("level", "error", "message", fmt.Sprintf("failed to sync time with ntp server %#q", ntpServer), "stack", microerror.JSON(err))
```

File: ntp/ntp.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	prometheus.MustRegister(syncErrorCount)
+
+	collector := &Collector{
+		logger: config.Logger,
+
+		ntpServers: config.NTPServers,
+
+		latencyHistogramVec: latencyHistogramVec,
+		latencyHistogramDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, "", "latency_seconds"),
```
Comment (kopiczko) [inline ntp/ntp.go:87]: Should we measure this in seconds?

File: helm/net-exporter/templates/daemonset.yaml (hunk 1 above)
Comment (corest) [reply helm/net-exporter/templates/daemonset.yaml:63]: Of course, we do. All the onprem installations are using custom NTP servers.

File: main.go (hunk 2 above)
Comment (corest) [reply main.go:38]: It is go native `flag` package, there is no `StringSliceVar`

File: ntp/ntp.go (hunk 4 above)
Comment (corest) [reply ntp/ntp.go:78]: `init` inside `ntp` package? I'm just following same structure as in other collectors

File: ntp/ntp.go (hunk 5 above)
Comment (corest) [reply ntp/ntp.go:91]: Then whole ntp-exporter codebase doesn't follow fmt.

File: ntp/ntp.go (hunk 4 above)
Comment (kopiczko) [reply ntp/ntp.go:87]: Let's leave like it is then.

**Comment by JosephSalisbury** [comment]
Seconds is the SI base unit for time, so it's the convention

**Pull request merged**